File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys 
input = sys.stdin.readline

n, m = map(int, input().split())
lan = []
for i in range(n):
    lan.append(int(input()))
    
start, end = 1, max(lan)

while start <= end:
    mid = (start+end)//2
    cnt = 0
    for i in lan:
        cnt += i//mid
        
    if cnt < m:
        end = mid-1
    else:
        start = mid+1

print(end)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int K = Integer.parseInt(input[0]);
        int N = Integer.parseInt(input[1]);
        int[] lan = new int[K];
        long start = 1;
        long end = 0;
        long mid = 0;
        for (int k = 0; k < K; k++) {
            lan[k] = Integer.parseInt(br.readLine());
            end = Math.max(end, lan[k]);
        }

        while (start <= end) {
            mid = (start+end)/2;
            long cnt = 0;
            for (int i = 0; i < K; i++) {
                cnt += lan[i]/mid;
            }

            if (cnt >= N) {
                start = mid+1;
            } else {
                end = mid-1;
            }
        }
        System.out.println(end);
    }
}
